allennlp/nlp_api.py

```diff
@@ -0,0 +1,289 @@
+                raise ConfigurationError("Unrecognized fallback behavior: %s" % fallback_behavior)
+        embedders = self._embedders[name]
+
+        if isinstance(embedders, torch.nn.Module):
+            return embedders.forward(input_tensors)
+        assert len(embedders) == len(input_tensors), \
+                "Mismatch between number of TokenIndexers (%d) and number of embedding modules (%d) " \
+                % (len(embedders), len(input_tensors))
+        embedded_arrays = []
+
+        for embedder, input_array in zip(embedders, input_tensors):
+            embedded_arrays.append(embedder.forward(input_array))
+        if len(input_tensors) == 1:
+            return embedded_arrays[0]
+        else:
+            return torch.cat(embedded_arrays, -1)
+
+    def get_recurrent_layer(self,
+                            name: str = "default",
+                            fallback_behavior: str = "crash") -> torch.nn.Module:
+        """
+        This method abstracts away the decision of which particular recurrent layer you want to use
+        when building your model, and allows for easy experimentation with different recurrent
+        layers.  All we do is a simple re-direction, where you specify a recurrent layer by name
+        (e.g., "passage_encoder" or "question_encoder"), and we return the layer with that name
+        that was passed to the constructor of ``NlpApi``.  If there was no layer with the name
+        you gave, there are a few things we could do, determined by ``fallback_behavior``.
+
+        Parameters
+        ----------
+        name : ``str``, optional (default=``"default"``)
+            The name of the recurrent layer, e.g., "sentence_encoder".  Multiple calls to this
+            method with the same name will return the same recurrent layer.  If ``name`` is not a
+            key in the ``recurrent_layers`` passed to :func:`__init__`, the behavior is defined by
+            the ``fallback_behavior`` parameter.
+        fallback_behavior : ``str``, optional (default=``"crash"``)
+            Determines what to do when ``name`` is not a key in ``recurrent_layers``.  There are
+            three options:
+
+            - ``"crash"``: raise an error.  This is the default.  The intention is to help you find
+              bugs - if you specify a particular encoder name when constructing your model without
+              giving a fallback behavior, you probably wanted to use a particular recurrent layer,
+              so we crash if it is not provided.
+            - ``"new default layer"``: In this case, we return a new encoder created with the
+              ``default_recurrent_layer`` function passed to the constructor.  If this is your
+              fallback behavior, you will get a `new` layer by default when you give a new name.
+              You should use this for things like stacked recurrent layers.
+            - ``"use [name]``: In this case, we `reuse` the layer with the given name, essentially
```

Comment:
I think this should no longer be an option - now, I'm presuming this class will be used in the `__init__` method of building a model (or `_build_model`), setting required layers as attributes which will be used in `forward`. Therefore, if you are using this for a second time rather than using the attribute you have already set, you are probably doing something wrong. Does that make sense?